server/src/main/scala/org/http4s/server/middleware/authentication/DigestAuth.scala

```diff
@@ -22,25 +22,60 @@ package authentication
 import cats.Monad
 import cats.data.Kleisli
 import cats.data.NonEmptyList
+import cats.effect.Concurrent
 import cats.effect.Sync
+import cats.effect.Timer
 import cats.syntax.all._
 import org.http4s.crypto.Hash
-import org.http4s.crypto.unsafe.SecureRandom
 import org.http4s.headers._
 
-import java.math.BigInteger
-import java.util.Date
 import scala.concurrent.duration._
 
 /** Provides Digest Authentication from RFC 2617.
   */
 object DigestAuth {
 
+  @deprecated(
+    "AuthenticationStore is going away, in favor of explicit subclasses of AuthStore. PlainTextAuthStore maintains the previous, insecure behaviour, whereas Md5HashedAuthStore is the new advised implementation going forward.",
+    "0.22.13",
+  )
+  type AuthenticationStore[F[_], A] = String => F[Option[(A, String)]]
+
+  sealed trait AuthStore[F[_], A]
```

Comment:
Does this need to be sealed?  I see we match on them in `checkAuthParams`.  Would it be better if those cases were a method on this, to permit more pluggable implementations?  Or are plaintext and MD5 the two that are specified?

Reply:
Based on how I understand the spec, there are only two sensible implementations:

#### Plain-text passwords:

    md5($un:$realm:$pw):$nonce:$nc:$cnonce:$qop:md5($method:$uri)

#### Pre-hashed passwords:

    md5($un:$realm:$pw) -> ha1 in the db
                    |
             retrieve from db
                    v
                   $ha1:$nonce:$nc:$cnonce:$qop:md5($method:$uri)

The way I see it, by pushing more into subclasses of `AuthStore`, we'd need to expose all these parameters for dubious benefit. `ha1` is the only secret that needs to be exposed to user code

Reply:
Reviewing the spec again for another comment, this is actually called out explicitly [by the spec](https://datatracker.ietf.org/doc/html/rfc2617#section-3.3):

> Note that the HTTP server does not actually need to know the user's
> cleartext password. As long as H(A1) is available to the server, the
> validity of an Authorization header may be verified.

Reply:
> An optional header allows the server to specify the algorithm used to
   create the checksum or digest. By default the MD5 algorithm is used
   and that is the only algorithm described in this document.

It looks like it can be `md5`, `md5-sess`, or extensions that probably nobody understands.  I think it would be unpleasant to register and understand more for little gain.

Reply:
The need to add support for this stuff may motivate new contributors 😉